service/raft/raft_group0.cc

```diff
@@ -669,70 +669,8 @@ future<> raft_group0::finish_setup_after_join() {
-        // 1. decommission is called immediately after upgrade, before we notice that the Raft feature is enabled
-        // 2. we don't notice it during decommission - including the long streaming phase - even until we reach
-        //    the `leave_group0` step which is almost at the end of the procedure
-        // 3. some node did notice that the feature is enabled and started `upgrade_to_group0`, using us
-        //    as one of the peers for running group 0 discovery, before we entered LEFT state
-        //
-        // These conditions together give a very unlikely scenario. If it does happen the user can perform
-        // the manual recovery procedure for group 0.
-        group0_log.warn(
-            "leave_group0: Raft feature not enabled yet. Assuming that the cluster is partially upgraded"
-            " and skipping the step. However, if your already finished the rolling upgrade procedure,"
-            " this means we just haven't noticed it yet. The internal upgrade-to-raft procedure on other nodes"
-            " may get stuck (they may try to contact us during the procedure). If that happens, manual recovery"
-            " will be required. Consult the documentation for more details.");
-        // TODO: link to the docs
-        co_return;
-    }
-
-    if (upgrade_state != group0_upgrade_state::use_post_raft_procedures) {
-        // The feature is enabled but `upgrade_to_group0` did not finish yet.
-        // The upgrade procedure requires everyone to participate. In order to not block others
-        // from doing their upgrades, we'll wait until we finish our procedure - then we can safely leave.
-        //
-        // FIXME: well, to be completely precise, it could happen that everyone enters `synchronize` state,
-        // then we're the first to enter `use_post_raft_procedures` state and consider the procedure finished,
-        // then we leave before others contact us to confirm that everyone entered `synchronize` and
-        // they get stuck trying to contact us. To completely ensure liveness we should wait not only until
-        // we enter `use_post_raft_procedures`, but until somebody else also does (then others can use that
-        // node to confirm that they can also finish their procedures).
-        // Extend `wait_until_upgrade_to_group0_finishes` with an extra step which RPCs a peer
-        // to confirm they entered `use_post_raft_procedures`.
-        group0_log.info("leave_group0: waiting until cluster fully upgrades to use Raft before proceeding...");
-        co_await _client.wait_until_group0_upgraded(_abort_source);
-        group0_log.info("leave_group0: cluster finished upgrade, continuing.");
-    }
-
-    // We're fully upgraded, we must have joined group 0.
-    if (!joined_group0()) {
-        on_internal_error(group0_log,
-            "leave_group0: we're fully upgraded to use Raft but didn't join group 0. Please report a bug.");
+    if (!(co_await raft_upgraded())) {
```

Comment:
Sorry to bikeshed but I think English is playing a nasty trick here. The name is ambiguous - are we waiting for the non-raft cluster to upgrade to raft, or a raft cluster to upgrade to a better raft ? The name implies the second. I think raft_upgrade_complete() is less ambiguous, or upgrade_to_raft_complete(), or something along these lines.

Reply:
Renamed to `raft_upgrade_complete`.